### PR TITLE
Full Windows Taskcluster CI

### PR DESCRIPTION
**Pull request by strseb**
This superseeds the PR: https://github.com/mozilla-mobile/mozilla-vpn-client/pull/2668

Full Taskcluster CI :
-> Has all Toolkit Changes from https://github.com/mozilla-mobile/mozilla-vpn-client/pull/2668
-> Now includes a hash-based index. (So we will recompile qt, when the script for the job changes)
-> in [f762244](https://github.com/mozilla-mobile/mozilla-vpn-client/pull/2893/commits/f7622447bb53ab7f63087f09569251aa4967228e) I applied fixes from @mkestler-mozilla to make the client build & work on qt6
-> Also ofc now has a job to build the final client for windows :) 

Example job: https://firefox-ci-tc.services.mozilla.com/tasks/bXWAkje6S0iVE88pf0YFfQ#artifacts

### CODE REVIEW

File: src/ui/main.qml [hunk 1]
```diff
@@ -6,6 +6,8 @@ import QtQuick 2.5
 import QtQuick.Controls 2.14
 import QtQuick.Layouts 1.14
 import QtQuick.Window 2.12
+import Qt5Compat.GraphicalEffects
```
Comment (strseb) [inline src/ui/main.qml:9]: This line seems to break qt5 compile support ... but without it the windows+qt6 client will crash 😅

Comment (mbirghan) [reply src/ui/main.qml:9]: Could we create a build variable just like MVPN_IOS and others for Qt5 and Qt6?

Comment (mkestler-mozilla) [reply src/ui/main.qml:9]: We agreed it is fine to break Qt5.

**Review by mbirghan — APPROVED** [review]

File: nebula/ui/components/VPNConnectionInfo.qml [hunk 2]
```diff
@@ -3,7 +3,7 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 import QtQuick 2.5
-import QtCharts 2.0
+import QtCharts 2.2
```
Comment (bakulf) [inline nebula/ui/components/VPNConnectionInfo.qml:6]: revert this change. VPNConnectionInfo is now totally different.

File: scripts/windows_compile.bat [hunk 3]
```diff
@@ -97,26 +97,25 @@ if "%WEBEXTENSION_BUILD%" == "T" (
...
-CALL :CopyDependency libssl-1_1-x64.dll %BUILDDIR%\bin\libssl-1_1-x64.dll
-CALL :CopyDependency libEGL.dll %BUILDDIR%\bin\libEGL.dll
-CALL :CopyDependency libGLESv2.dll %BUILDDIR%\bin\libGLESv2.dll
-CALL :CopyDependency Microsoft_VC142_CRT_x86.msm "%VCToolsRedistDir%\\MergeModules\\Microsoft_VC142_CRT_x86.msm"
-CALL :CopyDependency Microsoft_VC142_CRT_x64.msm "%VCToolsRedistDir%\\MergeModules\\Microsoft_VC142_CRT_x64.msm"
+CALL :CopyDependency libcrypto-1_1-x64.dll %BUILDDIR%\SSL\bin\libcrypto-1_1-x64.dll
+CALL :CopyDependency libssl-1_1-x64.dll %BUILDDIR%\SSL\bin\libssl-1_1-x64.dll
+
+ECHO "Checking vctools in %VCToolsRedistDir%"
+CALL :CopyDependency Microsoft_VC143_CRT_x86.msm "%VCToolsRedistDir%\\MergeModules\\Microsoft_VC143_CRT_x86.msm"
```
Comment (bakulf) [inline scripts/windows_compile.bat:111]: This is a big change... Do we really need VC143?

File: scripts/windows_compile.bat [hunk 4]
```diff
@@ -126,8 +125,7 @@ IF %DEBUG_BUILD%==T (
 
   cargo build --debug
   IF %ERRORLEVEL% NEQ 0 (
-    ECHO cargo failed for the extension!
```
Comment (bakulf) [inline scripts/windows_compile.bat:129]: revert this :)

File: scripts/windows_compile.bat [hunk 5]
```diff
@@ -140,8 +138,7 @@ IF %DEBUG_BUILD%==F (
 
   cargo build --release
   IF %ERRORLEVEL% NEQ 0 (
-    ECHO cargo failed for the extension!
-    EXIT 1
```
Comment (bakulf) [inline scripts/windows_compile.bat:144]: and this

File: scripts/windows_compile.bat [hunk 6]
```diff
@@ -152,23 +149,19 @@ ECHO Creating the project with flags: %FLAGS%
 
 if %DEBUG_BUILD% == T (
   ECHO Generating Debug Project
-  qmake -tp vc src/src.pro CONFIG+=debug %FLAGS%
+  qmake src/src.pro CONFIG+=debug %FLAGS%
```
Comment (bakulf) [inline scripts/windows_compile.bat:152]: why do we need to change how we compile the project? I'm OK with this, I just need to know more.

File: src/ui/main.qml (hunk 1 above)
Comment (bakulf) [reply src/ui/main.qml:9]: This is just a temporary fix of a bigger issue. Yes, the app starts, but it will crash in the following steps:
- opening the dev menu -> qtcharts2 plugin not found
- completing the auth -> graphical effects plugin not found
- on the main view -> localstorage plugin not found

**Review by bakulf — CHANGES_REQUESTED** [review]

File: scripts/windows_compile.bat (hunk 6 above)
Comment (mkestler-mozilla) [reply scripts/windows_compile.bat:152]: We should leave the -tp vc in.  For local developers this switch is what generates the Visual Studio project.

Comment (strseb) [reply scripts/windows_compile.bat:152]: We can't use msvc on the tc-windows vm as it's missing some .net dependency. (I can't resolve it, as we don't have admin) 
How about -vc as default and having a switch for nmake?

File: scripts/windows_compile.bat (hunk 3 above)
Comment (strseb) [reply scripts/windows_compile.bat:111]: I think we _could_ v142 with an older version of msvc +vsstudio. 
Whole reason for this is, as qt6 needed a newer winsdk to compile, thus I upgraded my local vsstudio to vs2022 at the same time and [it does not seem to offer VC142  anymore](https://docs.microsoft.com/de-de/visualstudio/install/workload-component-id-vs-community?view=vs-2022). (In fetches/generate_vsstudio_bundle.ps1 you can check what we have on the vm) 
I did not do a check of "what is the lowest possible thing here" - as I didn't think of this to be a major change. Can you elaborate on this? (Also in case you got qt6-static-win building with vs2017, you can just get me your sdk-version and i'll update the dev-bundle)

File: src/ui/main.qml (hunk 1 above)
Comment (bakulf) [reply src/ui/main.qml]: See the latest `main` commits. This is now fixed.

File: scripts/windows_compile.bat [hunk 7]
```diff
@@ -54,6 +51,14 @@ SET DEBUG_BUILD=F
 if "%1"== "--debug" SET DEBUG_BUILD=T
 if "%2"== "--debug" SET DEBUG_BUILD=T
 
+
+SET PROJECT_TYPE=vc
+if "%1"== "--nmake" SET PROJECT_TYPE=app
```
Comment (bakulf) [inline scripts/windows_compile.bat:56]: this is horrible. I'm sure there is a better way :)

File: taskcluster/scripts/build/windows.ps1 [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+./scripts/windows_compile.bat -w --nmake
+
+Write-Output "Writing Artifacts"
+
+New-Item -Path $REPO_ROOT_PATH/artifacts -ItemType "directory"
+Copy-Item -Path windows/installer/x64/MozillaVPN.msi -Destination ./artifacts/MozillaVPN.msi
+
+New-Item -Path $REPO_ROOT_PATH/unsigned -ItemType "directory"
+Copy-Item -Path libssl-1_1-x64.dll -Destination $REPO_ROOT_PATH/unsigned
```
Comment (bakulf) [inline taskcluster/scripts/build/windows.ps1:45]: can we use 'make install' for this?

File: taskcluster/scripts/build_qt/compile_qt_6.ps1 [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+$REPO_ROOT_PATH =resolve-path "$PSScriptRoot/../../../"
+$FETCHES_PATH =resolve-path "$REPO_ROOT_PATH/../../fetches"
+
+
+$BIN_PATH = "$REPO_ROOT_PATH/bin"
+
+Set-Location $FETCHES_PATH 
+Invoke-WebRequest -Uri https://download.qt.io/archive/qt/6.2/6.2.3/single/qt-everywhere-src-6.2.3.zip -OutFile qt-everywhere-src-6.2.3.zip
```
Comment (bakulf) [inline taskcluster/scripts/build_qt/compile_qt_6.ps1:12]: can you set a variable to pick the right version? we are migrating to 6.2.4 everywhere.

**Review by bakulf — CHANGES_REQUESTED** [review]

File: taskcluster/scripts/build/windows.ps1 (hunk 8 above)
Comment (bakulf) [reply taskcluster/scripts/build/windows.ps1:45]: I wonder if we can use something similar to this: https://github.com/mozilla-mobile/mozilla-vpn-client/blob/main/src/qmake/platforms/linux.pri#L109-L111

**Review by bakulf — APPROVED** [review]

File: taskcluster/scripts/build_qt/compile_qt_6.ps1 (hunk 9 above)
Comment (bakulf) [reply taskcluster/scripts/build_qt/compile_qt_6.ps1]: This is not done yet.

File: taskcluster/ci/build-qt/windows.yml [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    treeherder:
+        kind: build
+        platform: toolchains/opt
+        tier: 1
+
+windows:
+    description: "Windows QT compile Task"
+    index:
+        product: qt
+        job-name: qt-v6.2.3-windows
```
Comment (bakulf) [inline taskcluster/ci/build-qt/windows.yml:15]: maybe remove 6.2.3 from the name.

File: src/qmake/platforms/windows.pri [hunk 11]
```diff
@@ -89,3 +89,66 @@ HEADERS += \
...
+mozillavpnjson.path = $$PWD/../../../unsigned/
+INSTALLS += mozillavpnjson
+
+mozillavpnnp.files = $$PWD/../../../mozillavpnnp.exe
+mozillavpnnp.path = $$PWD/../../../unsigned/
+INSTALLS += mozillavpnnp
+
+
+
+
```
Comment (bakulf) [inline src/qmake/platforms/windows.pri:154]: extra lines here :)

File: src/qmake/platforms/windows.pri [hunk 12]
```diff
@@ -89,3 +89,66 @@ HEADERS += \
     platforms/windows/windowsstartatbootwatcher.h \
     systemtraynotificationhandler.h \
     wgquickprocess.h
+
+
+# Installs for the /unsigned folder
+
+mvpnExe.files = $$PWD/../../../mozillavpn.exe
+mvpnExe.path = $$PWD/../../../unsigned/
```
Comment (bakulf) [inline src/qmake/platforms/windows.pri:97]: I guess this doesn't work. you need to add:
`mvpnExe.CONFIG = no_check_exist executable`

File: src/qmake/platforms/windows.pri [hunk 13]
```diff
@@ -89,3 +89,66 @@ HEADERS += \
...
+INSTALLS += mvpnExe
+
+MozillaVPNpdb.files = $$PWD/../../../MozillaVPN.pdb
+MozillaVPNpdb.path = $$PWD/../../../unsigned/
```
Comment (bakulf) [inline src/qmake/platforms/windows.pri:101]: same here

File: src/qmake/platforms/windows.pri [hunk 14]
```diff
@@ -89,3 +89,66 @@ HEADERS += \
...
+VC142.files = $$PWD/../../../Microsoft_VC142_CRT_x64.msm
+VC142.path = $$PWD/../../../unsigned/
+INSTALLS += VC142
+
+tunnel.files = $$PWD/../../../windows/tunnel/x64/tunnel.dll
+tunnel.path = $$PWD/../../../unsigned/
+INSTALLS += tunnel
+
+wintun.files = $$PWD/../../../windows/tunnel/.deps/wintun/bin/amd64/wintun.dll
+wintun.path = $$PWD/../../../unsigned/
```
Comment (bakulf) [inline src/qmake/platforms/windows.pri:121]: and here

File: src/qmake/platforms/windows.pri [hunk 15]
```diff
@@ -89,3 +89,66 @@ HEADERS += \
...
+INSTALLS += wintun
+
+balrog.files = $$PWD/../../../balrog/x64/balrog.dll
```
Comment (bakulf) [inline src/qmake/platforms/windows.pri:124]: and here

File: src/qmake/platforms/windows.pri [hunk 16]
```diff
@@ -89,3 +89,66 @@ HEADERS += \
...
+
+WdfCoinstaller01011.files = $$PWD/../../../windows/split-tunnel/WdfCoinstaller01011.dll
+WdfCoinstaller01011.path = $$PWD/../../../unsigned/
+INSTALLS += WdfCoinstaller01011
+
+mozillavpnjson.files = $$PWD/../../../extension/manifests/windows/mozillavpn.json
+mozillavpnjson.path = $$PWD/../../../unsigned/
+INSTALLS += mozillavpnjson
+
+mozillavpnnp.files = $$PWD/../../../mozillavpnnp.exe
```
Comment (bakulf) [inline src/qmake/platforms/windows.pri:148]: and here

**Review by bakulf — CHANGES_REQUESTED** [review]

**Review by bakulf — APPROVED** [review]
